### PR TITLE
Add deployment

### LINKED ISSUE #2: Add fallbacks to content negotiation
The lambda returns the correct file eg. xml, jsonld, but it doesn't have a default so eg. /registry without an Accepts header just returns a 404. We'll need to change it so that html is the default, which means we'll also need to add a "does the file exist" check

### PR DESCRIPTION
**Pull request by kipparker**
Add content negotiation with a CDK cloudfront deployment

Fixes #2

### CODE REVIEW

File: script/ruby/ib1-registry.rb [hunk 1]
```diff
@@ -44,7 +44,11 @@ module Jena
   puts "Removing old output files..."
   FileUtils.rm_rf OUTPUT_DIR
```
Comment (kipparker) [inline script/ruby/ib1-registry.rb:46]: @bensummers I added a .gitignore file to the output folder to fix an issue with building on github, but it gets deleted by this command - is there a way of altering this to delete everything but .gitignore?

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        uses: actions/checkout@v3
+      - name: Install AWS CDK
+        run: npm install -g aws-cdk
+
+      - name: Build static site
+        run: |
+          docker build -t builder .
+          ls -al ${{ github.workspace }}
+          docker run --rm -v ${{ github.workspace }}:/code builder bash -c "pwd && ls -al"
+          docker run --rm -v ${{ github.workspace }}/output:/code/output builder bash -c "script/ib1-registry production ib1-root-registry"
```
Comment (kipparker) [inline .github/workflows/build.yml:31]: Build the files with docker - the GitHub runner environment has an older version of java that causes a conflict

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        run: aws s3 sync ./output s3://contentnegotiationstack-sitebucket397a1860-n4pplyx0o81v --delete
+      - uses: "actions/setup-python@v4"
+        with:
+          python-version: "3.12"
+      - name: Deploy CDK Stack
+        run: |
+          cd content-negotiation
+          pip install pipenv
+          pipenv install --system --deploy --ignore-pipfile 
+          cdk deploy --all --require-approval never
```
Comment (kipparker) [inline .github/workflows/build.yml]: Run CDK deploy to make any infrastructure changes

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+      - name: Invalidate Cache
+        run: aws cloudfront create-invalidation --distribution-id E2OLIP3E7LAQ2F --paths "/*"
```
Comment (kipparker) [inline .github/workflows/build.yml]: Invalidate cache

File: content-negotiation/content_negotiation/content_negotiation_stack.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    aws_cloudfront as cloudfront,
+    aws_cloudfront_origins as origins,
+    aws_lambda as _lambda,
+    aws_iam as iam,
+    RemovalPolicy,
+)
+from constructs import Construct
+
+
+class ContentNegotiationStack(Stack):
```
Comment (kipparker) [inline content-negotiation/content_negotiation/content_negotiation_stack.py:15]: Create the s3 bucket and cloud front distribution

File: content-negotiation/content_negotiation/lambda_edge_alias_stack.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+    def __init__(self, scope: Construct, construct_id: str, **kwargs) -> None:
+        super().__init__(scope, construct_id, **kwargs)
+        lambda_edge_role = iam.Role(
+            self,
+            "LambdaEdgeRole",
+            assumed_by=iam.ServicePrincipal("lambda.amazonaws.com"),
+            managed_policies=[
+                iam.ManagedPolicy.from_aws_managed_policy_name(
+                    "service-role/AWSLambdaBasicExecutionRole"
```
Comment (kipparker) [inline content-negotiation/content_negotiation/lambda_edge_alias_stack.py]: Lambda@edge stack is separate as it has to deploy into us-east-1

File: content-negotiation/lambda/index.py [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+        accept = headers["accept"][0]["value"]
+        if "application/ld+json" in accept:
+            uri += ".jsonld"
+        elif "text/turtle" in accept:
+            uri += ".ttl"
+        elif "application/rdf+xml" in accept:
+            uri += ".rdf"
+
+    request["uri"] = uri
+    return request
```
Comment (kipparker) [inline content-negotiation/lambda/index.py]: Content negotiation logic

File: script/ruby/ib1-registry.rb [hunk 8]
```diff
@@ -44,7 +44,11 @@ module Jena
...
 end
-FileUtils.mkdir OUTPUT_DIR
+begin
+  FileUtils.mkdir OUTPUT_DIR
+rescue Errno::EEXIST
+  # ignore
+end
```
Comment (kipparker) [inline script/ruby/ib1-registry.rb:52]: Solves an issue when running on GitHub environment - mkdir was failing with Errno:EEXIST, I guess FileUtils.rm_rf failed due to permissions or something

File: script/ruby/ib1-registry.rb (hunk 1 above)
Comment (bensummers) [reply script/ruby/ib1-registry.rb:46]: You solution below looks OK, but moving the entry from the root .gitignore to this one will mean it gets wiped out every time it's run in development, and so easy to accidentally commit the output. I'll address this.

File: content-negotiation/lambda/index.py (hunk 7 above)
Comment (bensummers) [reply content-negotiation/lambda/index.py]: This needs a few tweaks:
- Fallback to HTML version when no valid format requested, or a browser requests the URL
- Need to set response Content-Type for JSONLD and Turtle files (XML version has right MIME type)
- Response needs "Vary: Accept" header.

File: script/ruby/ib1-registry.rb (hunk 1 above)
Comment (kipparker) [reply script/ruby/ib1-registry.rb:46]: Moving the .gitignore file here is a way of committing an empty folder so it's available for the build. I got the build working by mounting a volume to /output so I can use it from outside the docker container to push to s3, but the folder needs to exist for that to work

File: content-negotiation/lambda_code/index.py [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+        file_extension = ".jsonld"
+    elif "text/turtle" in accept_header:
+        content_type = "text/turtle"
+        file_extension = ".ttl"
+    elif "application/xml" in accept_header or "text/xml" in accept_header:
+        content_type = "application/xml"
+        file_extension = ".xml"
+    else:
+        content_type = "text/html"
+        file_extension = ".html"
```
Comment (kipparker) [inline content-negotiation/lambda_code/index.py:41]: Adds an HTML fallback

File: content-negotiation/lambda_code/index.py [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+
+    # Modify the URI to point to the correct file extension
+    request["uri"] = request["uri"] + file_extension
+
+    # Add headers to the request object
+    request["headers"]["content-type"] = [
+        {"key": "Content-Type", "value": content_type}
+    ]
```
Comment (kipparker) [inline content-negotiation/lambda_code/index.py:49]: Set the content type header

File: content-negotiation/content_negotiation/lambda_edge_alias_stack.py [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+            ],
+        )
+
+        self.lambda_edge_function = _lambda.Function(
+            self,
+            "LambdaContentNegotiation",
+            runtime=_lambda.Runtime.PYTHON_3_12,
+            handler="index.handler",
+            code=_lambda.Code.from_asset("lambda_code"),
+            role=lambda_edge_role,
```
Comment (kipparker) [inline content-negotiation/content_negotiation/lambda_edge_alias_stack.py]: Changed the location of the lambda code, `lambda` being a reserved word

**Comment by kipparker** [comment]
@bensummers think it might be a good idea to add https://github.com/icebreakerone/registry-core as another folder in here rather than a separate repo? Then we can easily rebuild both if/when the registry code changes

File: content-negotiation/lambda_code/index.py [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+    return bool(re.search(pattern, text))
+
+
+def handler(event, context):
+    # Extract the request details
+    request = event["Records"][0]["cf"]["request"]
+    headers = request["headers"]
+    print(f"Request URI: {request['uri']}")
+    print(
+        f"Accept header: {headers.get('accept', [{'value': 'text/html'}])[0]['value']}"
```
Comment (kipparker) [inline content-negotiation/lambda_code/index.py:19]: Logging for the lambda functino

File: content-negotiation/lambda_code/index.py [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+    )
+    # Default to HTML if no Accept header is present
+    accept_header = headers.get("accept", [{"value": "text/html"}])[0]["value"]
+    # If the file extension in request['uri'] is present, return the request as is
+    if _has_extension(request["uri"]):
+        return request
```
Comment (kipparker) [inline content-negotiation/lambda_code/index.py:25]: If the extension is supplied, we return the request - allows direct requests for .html or .css files

File: content-negotiation/lambda_code/index.py [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+    if request["uri"] == "/":
+        # Return the index page with HTML content type
+        request["uri"] = "/index"
```
Comment (kipparker) [inline content-negotiation/lambda_code/index.py:28]: As we're messing around with paths, we need to deal with the root page / directly

**Review by kipparker — COMMENTED** [review]
@bensummers I've made some changes to the logic, and had to fix the deployment which is working now - the url is https://d2jum2j743d83t.cloudfront.net/

**Pull request merged**